apps/language_server/lib/language_server/experimental/provider/handlers/goto_definition.ex

```diff
@@ -0,0 +1,27 @@
+defmodule ElixirLS.LanguageServer.Experimental.Provider.Handlers.GotoDefinition do
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Requests.GotoDefinition
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Responses
+  alias ElixirLS.LanguageServer.Experimental.SourceFile
+  alias ElixirLS.LanguageServer.Experimental.SourceFile.Conversions
+  require Logger
+
+  def handle(%GotoDefinition{} = request, _) do
+    source_file = request.source_file
+    pos = request.position
+
+    source_file_string = SourceFile.to_string(source_file)
+
+    with %ElixirSense.Location{} = location <-
+           ElixirSense.definition(source_file_string, pos.line, pos.character + 1),
+         {:ok, lsp_location} <- Conversions.to_lsp(location, source_file) do
+      {:reply, Responses.GotoDefinition.new(request.id, lsp_location)}
+    else
+      nil ->
+        {:reply, Responses.GotoDefinition.new(request.id, nil)}
+
+      {:error, reason} ->
+        Logger.error("GotoDefinition failed: #{inspect(reason)}")
+        {:error, Responses.GotoDefinition.error(request.id, :request_failed, inspect(reason))}
```

Comment:
you're going to hate me, but someone I respect a great deal got me in the habit of "with statements should have only zero or one else clause". Is there a way we can tighten this up?
perhaps by making a function like:
```elixir
defp to_responase(request_id, %ElixirSense.Location{} = location, %SourceFile{} = source_file) do 
  with {:ok, lsp_location} <- Conversions.to_lsp(location, source_file) do 
    {:ok, Responses.GotoDefinition.new(request_id, lsp_location)}
   end
end

defp to_response(request_id, nil, _source_file)  do
 {:ok, Responses.GoToDefinition.new(request_id, nil)}
end
```
Then the with is 
```elixir
with {:ok, maybe_location} <- ElixirSense.definition(source_file_string, pos.line, pos.character + 1),
     {:ok, response} <- to_response(request.id, location, source_file) do 
  {:reply, response}
else error -> 
    ...
 end
```

Reply:
First, I agree to break the 'else clause into two functions here.

The main reason is that these else clauses come from different step logic, one ElixirSense requested result and one conversion.

Also, I learned about this rule from [here](https://github.com/lucasvegi/Elixir-Code-Smells#complex-else-clauses-in-with) a long time ago, but I don't think this rule can be applied to all scenarios, for example, in some main functions (the function calling it is not transparent), sometimes we don't need to follow this rule. Just like in oban's custom [process](https://getoban.pro/docs/pro/0.13.1/Oban.Pro.Worker.html#c:process/1) callback, if you need to use different retry strategies according to different 'else' clauses, it would be more cohesive to put them together.